x/operator/keeper/state_update.go

```diff
@@ -0,0 +1,494 @@
+package keeper
+
+import (
+	"fmt"
+
+	errorsmod "cosmossdk.io/errors"
+	sdkmath "cosmossdk.io/math"
+
+	delegationtype "github.com/ExocoreNetwork/exocore/x/delegation/types"
+	"github.com/ExocoreNetwork/exocore/x/operator/types"
+	types2 "github.com/ExocoreNetwork/exocore/x/restaking_assets_manage/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+type AssetPriceAndDecimal struct {
+	Price        sdkmath.Int
+	PriceDecimal uint8
+	Decimal      uint32
+}
+
+type slashAmounts struct {
+	AmountFromUnbonding sdkmath.Int
+	AmountFromOptedIn   sdkmath.Int
+}
+type SlashAssets struct {
+	slashStakerInfo   map[string]map[string]*slashAmounts
+	slashOperatorInfo map[string]*slashAmounts
+}
+
+func (k *Keeper) UpdateOptedInAssetsState(ctx sdk.Context, stakerID, assetID, operatorAddr string, opAmount sdkmath.Int) error {
+	// get the AVS opted-in by the operator
+	avsList, err := k.GetOptedInAVSForOperator(ctx, operatorAddr)
+	if err != nil {
+		return err
+	}
+	// get price and priceDecimal from oracle
+	price, decimal, err := k.oracleKeeper.GetSpecifiedAssetsPrice(ctx, assetID)
+	if err != nil {
+		return err
+	}
+
+	// get the decimal of asset
+	assetInfo, err := k.restakingStateKeeper.GetStakingAssetInfo(ctx, assetID)
+	if err != nil {
+		return err
+	}
+	opUSDValue := CalculateShare(opAmount, price, assetInfo.AssetBasicInfo.Decimals, decimal)
+	for _, avs := range avsList {
+		// get the assets supported by the AVS
+		avsSupportedAssets, err := k.avsKeeper.GetAvsSupportedAssets(ctx, avs)
+		if err != nil {
+			return err
+		}
+
+		if _, ok := avsSupportedAssets[assetID]; ok {
+			// UpdateStakerShare
+			err = k.UpdateStakerShare(ctx, avs, stakerID, operatorAddr, opUSDValue)
+			if err != nil {
+				return err
+			}
+
+			// UpdateStateForAsset
+			changeState := types.OptedInAssetState{
+				Amount: opAmount,
+				Value:  opUSDValue,
+			}
+			err = k.UpdateStateForAsset(ctx, assetID, avs, operatorAddr, changeState)
+			if err != nil {
+				return err
+			}
+
+			// UpdateOperatorShare
+			err = k.UpdateOperatorShare(ctx, avs, operatorAddr, opUSDValue)
+			if err != nil {
+				return err
+			}
+
+			// UpdateAVSShare
+			err = k.UpdateAVSShare(ctx, avs, opUSDValue)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
```

Comment:
A lot of information is being stored by this function. Where is it used? Is all of it required?

Reply:
In the operator module, the USD share states related to assets, stakers, operators, and AVS will be stored separately. This function will update the USD share states related to assets, stakers, operators, and AVS when the asset amount changes due to delegation, undelegation, slashStaker, and slashOperator. In the current implementation, the stored states haven't been uniformly optimized, so all of them are required

Reply:
I think we should work on this design.

Reply:
Yeah, the data structure of the stored information in the assets, delegation and operator modules should be redesigned and uniformly optimized. Maybe we can work on it after all the basic features have been finished.